// This patch is for https://github.com/openxla/xla/commit/ec0177de1748b4ebb0ecbd6f26043fdb1eb47d24.
// It can be removed in the next openXLA pin update after 01/26/2024.
diff --git a/xla/service/gpu/gpu_executable.cc b/xla/service/gpu/gpu_executable.cc
index 0f1818be2..c181f3025 100644
--- a/xla/service/gpu/gpu_executable.cc
+++ b/xla/service/gpu/gpu_executable.cc
@@ -382,9 +382,13 @@ absl::Status ExecuteThunks(const std::string& module_name,
     }
   }
 
-  // Maybe join a round of rendezvous after thunk initialization.
-  TF_RETURN_IF_ERROR(
-      MaybeRendezvousAfterInitialization(run_options, thunks_initialized));
+  // Maybe join a round of rendezvous after thunk initialization. We do this
+  // only in presence of collective cliques which means that we have collective
+  // operations in the XLA operations that tend to cause deadlocks.
+  if (!collective_cliques.empty()) {
+    TF_RETURN_IF_ERROR(
+        MaybeRendezvousAfterInitialization(run_options, thunks_initialized));
+  }
 
   // Prepare parameters for thunks execution.
   Thunk::ExecuteParams execute_params = Thunk::ExecuteParams::Create(
diff --git a/xla/service/gpu/thunk.h b/xla/service/gpu/thunk.h
index 51a566b8f..94bab421f 100644
--- a/xla/service/gpu/thunk.h
+++ b/xla/service/gpu/thunk.h
@@ -175,6 +175,8 @@ class Thunk {
     absl::StatusOr<NcclComm::Lock> GetComm(const NcclCliqueKey& clique_key,
                                            int32_t rank) const;
 
+    bool empty() const { return cliques_map_.empty(); }
+
    private:
     CliquesMap cliques_map_;
   };
